### PR TITLE
executor,table: support replace operation for table partition

### PR DESCRIPTION
**Pull request by tiancaiamao**
## What have you changed? (mandatory)

Some modify on batchChecker to make it support partitioned table.

* Add a `t table.Table` field into `toBeCheckedRow`, because the old representation does not have enough information about which partition it belongs to.
* Add a `GetPartitionByRow` method to the table.PartitionedTable interface.
* Tiny modify `locatePartition` function to support `NULL` value.

## What is the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

unit test

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
no

## Does this PR affect tidb-ansible update? (mandatory)
no

## Does this PR need to be added to the release notes? (mandatory)
no

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @jackysp @coocood

File: executor/write_test.go [hunk 1]
```diff
@@ -777,6 +777,145 @@ func (s *testSuite) TestReplace(c *C) {
...
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("set @@session.tidb_enable_table_partition=1")
+	testSQL := `drop table if exists replace_test;
+    create table replace_test (id int PRIMARY KEY AUTO_INCREMENT, c1 int, c2 int, c3 int default 1) partition by range (id) (
+		PARTITION p0 VALUES LESS THAN (3),
+		PARTITION p1 VALUES LESS THAN (5),
+		PARTITION p2 VALUES LESS THAN (7),
+		PARTITION p3 VALUES LESS THAN (9)
+);`
```
Comment (ciscoxll) [inline executor/write_test.go]: Please align `sql`.

File: executor/write_test.go [hunk 2]
```diff
@@ -777,6 +777,145 @@ func (s *testSuite) TestReplace(c *C) {
...
+	tk.MustExec("rollback")
+
+	tk.MustExec(`drop table if exists replace_test_1`)
+	tk.MustExec(`create table replace_test_1 (id int, c1 int) partition by range (id) (
+		PARTITION p0 VALUES LESS THAN (4),
+		PARTITION p1 VALUES LESS THAN (6),
+		PARTITION p2 VALUES LESS THAN (8),
+		PARTITION p3 VALUES LESS THAN (10),
+		PARTITION p4 VALUES LESS THAN (100)
+)`)
```
Comment (ciscoxll) [inline executor/write_test.go]: ditto.

File: table/tables/partition.go [hunk 3]
```diff
@@ -156,18 +156,27 @@ func partitionRecordKey(pid int64, handle int64) kv.Key {
...
+	var isNull bool
 	partitionExprs := t.partitionExpr.UpperBounds
 	idx := sort.Search(len(partitionExprs), func(i int) bool {
 		var ret int64
-		ret, _, err = partitionExprs[i].EvalInt(ctx, chunk.MutRowFromDatums(r).ToRow())
+		ret, isNull, err = partitionExprs[i].EvalInt(ctx, chunk.MutRowFromDatums(r).ToRow())
 		if err != nil {
 			return true // Break the search.
 		}
+		if isNull {
```
Comment (jackysp) [inline table/tables/partition.go]: Is it a bug fix?

Comment (tiancaiamao) [reply table/tables/partition.go]: Yes, a bug fix.
It's a known issue, I did not take special care to `NULL` and plan to do it later.
but I meet some problems in the test code, so I change my mind and fix it here.

Comment (zz-jason) [reply table/tables/partition.go]: do we need to cherrypick this bugfix to release-2.0?

Comment (jackysp) [reply table/tables/partition.go]: I think there is no plan to support partition in release-2.0.

**Comment by jackysp** [comment]
Please resolve the conflicts.

**Comment by tiancaiamao** [comment]
PTAL @coocood

File: executor/batch_checker.go [hunk 4]
```diff
@@ -48,12 +50,27 @@ type batchChecker struct {
 	dupOldRowValues map[string][]byte
 }
 
-// batchGetOldValues gets the values of storage in batch.
-func (b *batchChecker) batchGetOldValues(ctx sessionctx.Context, t table.Table, handles []int64) error {
-	batchKeys := make([]kv.Key, 0, len(handles))
-	for _, handle := range handles {
-		batchKeys = append(batchKeys, t.RecordKey(handle))
+func (b *batchChecker) RemoveRow(ctx sessionctx.Context, t table.Table, handle int64, row []types.Datum) error {
```
Comment (coocood) [inline executor/batch_checker.go]: Seems only used in `Replace`, why extract this function to `batchChecker` instead of `ReplaceExec`?

Comment (tiancaiamao) [reply executor/batch_checker.go]: I have planned to wrap a new API layer around `batchChecker`,  RemoveRow/AddRow/UpdateRow ...
The benefit of that API layer is that we won't be trapped by this kind of invoking:

```
// blah blah ...
table.RemoveRecord() // modify table
// forget to backfill batchChecker indices...
```

But I found it's too much changes to accomplish in this single PR.

Should I reset this change, or just keep it here and add other API later? @coocood

**Comment by tiancaiamao** [comment]
@coocood

File: executor/batch_checker.go (hunk 4 above)
Comment (coocood) [reply executor/batch_checker.go]: I think reset the change is better.

**Comment by tiancaiamao** [comment]
PTAL @coocood

**Comment by tiancaiamao** [comment]
PTAL @coocood

**Comment by coocood** [comment]
LGTM

**Pull request merged**